kube/src/api/metadata.rs

```diff
@@ -75,14 +75,44 @@ pub trait Resource {
 
     /// Metadata that all persisted resources must have
     fn meta(&self) -> &ObjectMeta;
+    /// Metadata that all persisted resources must have
+    fn meta_mut(&mut self) -> &mut ObjectMeta;
+}
+
+/// Helper methods for resources.
+pub trait ResourceExt: Resource {
     /// The name of the resource
-    fn name(&self) -> String;
+    fn name(&self) -> Option<String>;
+    /// Returns the name of the resource, panicking if it is
+    /// missing. Use this function if you know that name is set, for example
+    /// when resource was received from the apiserver.
+    /// Because of `.metadata.generateName` field, in other contexts name
+    /// may be missing.
+    fn expect_name(&self) -> String;
     /// The namespace the resource is in
     fn namespace(&self) -> Option<String>;
     /// The resource version
     fn resource_ver(&self) -> Option<String>;
 }
 
+impl<K: Resource> ResourceExt for K {
+    fn name(&self) -> Option<String> {
+        self.meta().name.clone()
+    }
+
+    fn expect_name(&self) -> String {
+        self.meta().name.clone().expect("kind has metadata.name")
+    }
+
+    fn namespace(&self) -> Option<String> {
+        self.meta().namespace.clone()
+    }
+
+    fn resource_ver(&self) -> Option<String> {
+        self.meta().resource_version.clone()
+    }
```

Comment:
I think we can have a lot more helpers in here if we are uncoupling the more foundational `Resource` trait, with a more user-friendly one.

This should be a helper trait for dealing with `ObjectMeta`, it needs to cover at least the basics: stuff like, labels, annotations, ownerrefs, finalizers, managedfields, where we can unpack optional vectors into empty vectors.

Reply:
I will add labels, annotations, ownerrefs, and annotations (getters and setters), and also getter for UID.

I don't think we should add managedfields to the trait - AFAIU one does not need to look at `.metadata.managedFields` or modify it unless one is writing an alternative apiserver implementation.

Reply:
ok, fair point on `managedFields`.

btw, i am not sure about setters, because most of the time people change them with `apply` patches, or more awkward json patches, and setters kind of encourage use of `api.replace` which is awkward, and can run into idempotency issues.

i think maybe some patchbuilder utils for changing metadata fields might be more appropriate.